### DISCUSSION TITLE
Dropdown in form filtered by another dropdwpn condition

### DISCUSSION THREAD

**Discussion by Emeronerp1** [serenity-is/Serenity#6628] [Help] [answered]
Hi I'm new to serenity.
I have four forms basically called as
Sales quotation, sales order, sales delivery and sales invoice.

In sales invoice I have a dropdown called against and there it will load against quotation, against order  against delivery.

If I click on against quotation I want to load all the quotations created in another dropdown, or if click on against sales order I want to load all created orders in the same dropdown, same applicable for sales delivery also.

I'm stuck here, I don't know how to create the field in salesinvoice row and lookupscript, because this lookupscript is linked with three different rows ie
Salesquoterow,salesorderrow,salesdeliveryrow.

Also I need to apply a condition on all this  , like if an invoice created against sales quotation it should not show in the dropdown, basically I need to load number, or invoice number 

Appreciate on your support.

Thanks

**Comment by minhhungit** [comment]
search with keyword "cascade"

https://serenity.is/Docs/howto/how_to_setup_cascaded_editors

**Comment by wezmag** [comment, marked as answer]
## There are two methods to solve your problem.

### Method 1:
Let's assume in the SalesInvoiceForm, there are two fields: `Against` (enum) and `AgainstId` (int).
`Against` field stores the selected type which can be Quotation, Order or Delivery.
`AgainstId` field stores the Id of the entity of the selected type. 
When `Against` is selected, we need to populate `AgainstId` dropdown with the entities of the selected type.

Before we start, we need to create an empty dropdown editor for `AgainstId`.
Create a new TypeScript file EmptyDropdownEditor.ts with content as below:
``` TypeScript
import { Decorators, Select2Editor } from "@serenity-is/corelib";

@Decorators.registerEditor("MyProject.MyModule.EmptyDropdownEditor")
export class EmptyDropdownEditor extends Select2Editor<any, any> {
    constructor(hidden: JQuery) {
        super(hidden);
    }
}
```
Build your project, and then apply `[EmptyDropdownEditor]` attribute to `AgainstId` in the SalesInvoiceForm.
When you run your project now, you should have `Against` and `AgainstId` fields in the form and both are dropdowns.

Add below code in SalesInvoiceDialog.ts
``` TypeScript
constructor() {
    super();
    
    // when Against changes, populate AgainstId dropdown
    this.form.Against.changeSelect2(e => {
        const selectedAgainst = parseInt(this.form.Against.value) as Against;  //enum
        this.populateAgainstId(selectedAgainst);
    });
    
    // when Against value is set in Edit mode, populate AgainstId dropdown
    this.form.AgainstType.setEditValue = (source, property) => {
        var val = source[property.name];
        this.form.AgainstType.set_value((val == null ? null : val.toString()));
        this.populateAgainstId(val as AgainstType);
    };
}

private populateAgainstId(against: Against): void {
    this.form.AgainstId.clearItems();
    switch (against) {
        case Against.Quotation:
            const quotations = QuotationRow.getLookup().items;
            quotations.forEach(item => {
                this.form.AgainstId.addItem(<Select2Item>{ id: item[QuotationRow.idProperty].toString(), text: item[QuotationRow.nameProperty] });
            })
            break;
        case Against.Order:
            const orders = OrderRow.getLookup().items;
            orders.forEach(item => {
                 this.form.AgainstId.addItem(<Select2Item>{ id: item[OrderRow.idProperty].toString(), text: item[OrderRow.nameProperty] });
            });
            break;
        case Against.Delivery:
            const deliveries = DeliveryRow.getLookup().items;
            deliveries.forEach(item => {
                 this.form.AgainstId.addItem(<Select2Item>{ id: item[DeliveryRow.idProperty].toString(), text: item[DeliveryRow.nameProperty] });
            });
            break;
        }
    }
}
```

### Method 2:
1. Create a view includes all entities from Quotation, Order and Delivery tables. The view should at least contain three columns, Against, Id and Text.
2. Generate a Row.cs from the view, apply `[LookupScript]` attribute to it.
3. Apply `[LookupEditor(typeof(NewlyCreatedRow))]` to AgainstId in SalesInvoiceForm.
4. Set up cascaded editor for Against and AgainstId.

**Reply by Emeronerp1** [reply]
Thank you somuch, it worked.
